Add dataframe row and column selections

lukasmasuch opened the pull request:
## Describe your changes

This PR adds row and column selection support to `st.dataframe`. It can be used like this:

```python
selection = st.dataframe(
    df,
    on_select="rerun",
    selection_mode="single-row"
)
```

- [Demo App](https://dataframe-row-selections.streamlit.app/)

https://github.com/streamlit/streamlit/assets/2852129/f3ff476a-0bd0-4b82-bc97-6bda3a3be98c

## GitHub Issue Link (if applicable)

- Closes https://github.com/streamlit/streamlit/issues/688
- Closes https://github.com/streamlit/streamlit/issues/7134
- https://github.com/streamlit/streamlit/issues/455
- https://github.com/streamlit/streamlit/issues/8319

## Testing Plan

- Added unit tests
- Added e2e tests (see https://github.com/streamlit/streamlit/pull/8634)

```[tasklist]
### e2e tests
- [x] Single row/column selection
- [x] Multi row/column selection
- [x] Mixed selections
- [x] Screenshot of a dataframe with multiple selections
- [x] Clear selections via toolbar
- [x] Clear selections via escape
- [x] Select all rows in multi-row selection via top checkbox
- [x] Optional: Test drag and drop selection
- [x] Optional: Test shift selections
- [x] Optional: Test selections in form
- [x] Have some test cases work with session state and others with return value
- [x] Add a test case validating that the callback gets called
``` 


---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch committed:
Add initial version of row selection

lukasmasuch committed:
Update selection state format

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into feature/dataframe-selections

lukasmasuch committed:
Fix signature

lukasmasuch committed:
Improve frontend handling

lukasmasuch committed:
Update to alpha 7

lukasmasuch committed:
Fix debounce

lukasmasuch committed:
Additional flag

lukasmasuch committed:
Prevent circular import

lukasmasuch committed:
Fix test

lukasmasuch committed:
Minor improvement

lukasmasuch committed:
Add dataframe selection demo

lukasmasuch committed:
Update docs and parameters

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into feature/dataframe-selections

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into feature/dataframe-selections

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into feature/dataframe-selections

lukasmasuch committed:
Add colum selection mode

lukasmasuch committed:
Fix issue

lukasmasuch committed:
Keep old selections

lukasmasuch committed:
Always use the square checkbox

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into feature/dataframe-selections

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Add cache replay check

lukasmasuch committed:
Minor dataframe related improvements

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Sorted list

lukasmasuch committed:
Apply some refactorings

lukasmasuch committed:
Add some tests

lukasmasuch committed:
Remove unused import

lukasmasuch committed:
Some more minor improvements

lukasmasuch committed:
Fix test

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Fix comment in proto

lukasmasuch committed:
Add more tests

lukasmasuch committed:
Fix comment

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/hooks/useDataLoader.ts:
```diff
@@ -82,7 +80,6 @@ function useDataLoader(
         const arrowCell = data.getCell(originalRow + 1, originalCol)
         return getCellFromArrow(column, arrowCell, data.cssStyles)
       } catch (error) {
-        logError(error)
```
Drive-by cleanup: this logging isn't really necessary and this case is also triggered ins situations where its totally ok that it runs into this issue (the data gets updated and the new data has less rows).

lukasmasuch committed:
Move import

lukasmasuch committed:
Add comment

lukasmasuch committed:
Change comment

LukasMasuch commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+      dataEditorRef.current?.updateCells(cells)
+    },
+    []
+  )
```
These callbacks were just moved, but not really modified

lukasmasuch committed:
Some more cleanup

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+  )
+
+  // This callback is used to clear all selections (row/column/cell)
+  const clearSelection = React.useCallback(() => {
+    const emptySelection = {
+      columns: CompactSelection.empty(),
+      rows: CompactSelection.empty(),
+      current: undefined,
+    }
+    setGridSelection(emptySelection)
+    applySelections(emptySelection)
+  }, [applySelections])
+
+  // This callback is used to clear only cell selections
+  const clearCellSelection = React.useCallback(() => {
+    setGridSelection({
+      columns: gridSelection.columns,
+      rows: gridSelection.rows,
+      current: undefined,
+    })
+  }, [gridSelection])
+
+  // This callback is used to refresh the rendering of specified cells
+  const refreshCells = React.useCallback(
+    (
+      cells: {
+        cell: [number, number]
+      }[]
+    ) => {
+      dataEditorRef.current?.updateCells(cells)
+    },
+    []
+  )
```
These callbacks just got moved without any logical changes

lukasmasuch committed:
Fix wrong check

lukasmasuch committed:
Fix test

raethlein commented on frontend/lib/package.json:
```diff
@@ -38,9 +38,9 @@
-    "@glideapps/glide-data-grid": "6.0.4-alpha3",
-    "@glideapps/glide-data-grid-cells": "6.0.4-alpha3",
-    "@glideapps/glide-data-grid-source": "6.0.4-alpha3",
+    "@glideapps/glide-data-grid": "6.0.4-alpha7",
+    "@glideapps/glide-data-grid-cells": "6.0.4-alpha7",
+    "@glideapps/glide-data-grid-source": "6.0.4-alpha7",
```
just out of curiosity: what is the release strategy of glide and is it safe to use alpha versions? Since you are a glide maintainer I assume yes 🙂

lukasmasuch replied:
Yep, we have been running on "alpha" releases for quite some time since we need some of the features & fixes in the latest state. But these alpha releases are pretty stable.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -78,7 +80,7 @@ import "@glideapps/glide-data-grid-cells/dist/index.css"
 
 // Debounce time for triggering a widget state update
 // This prevents to rapid updates to the widget state.
```
nit: This prevents rapid updates to the widget state.

lukasmasuch replied:
Fixed

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+      // into a single function that updates the widget state with both the editing
+      // state and the selection state.
+
+      const selectionState = {
+        // We use snake case here since this is the widget state
+        // that is sent and used in the backend. Therefore, it should
+        // conform with the Python naming conventions.
+        select: {
+          rows: [] as number[],
+          columns: [] as string[],
+        },
+      }
+
+      selectionState.select.rows = newSelection.rows.toArray().map(row => {
+        return getOriginalIndex(row)
+      })
+      selectionState.select.columns = newSelection.columns
+        .toArray()
+        .map(columnIdx => {
+          return getColumnName(originalColumns[columnIdx])
+        })
+      const newWidgetState = JSON.stringify(selectionState)
```
suggestion: I would move this part to a function outside of the React component, as it does not use anything of it.

```typescript
function getSelectionState(newSelection: GridSelection, getIndexForRow: (row: Row) => index, getColumnName: (column: Column) => string): GridSelection {
        const selectionState = {
        // We use snake case here since this is the widget state
        // that is sent and used in the backend. Therefore, it should
        // conform with the Python naming conventions.
        select: {
          rows: [] as number[],
          columns: [] as string[],
        },
      }

      selectionState.select.rows = newSelection.rows.toArray().map(row => {
        return getIndexForRow(row)
      })
      selectionState.select.columns = newSelection.columns
        .toArray()
        .map(columnIdx => {
          return getColumnName(originalColumns[columnIdx])
        })
      return selectionState;
}
```

and then call that function from here and passing the component-dependent information. This way only the component-specific parts are in here and the logic can live standalone. In general this should make it easier to understand a components-flow by reducing the number of lines within a component and clearly seeing where an update is triggered.

lukasmasuch replied:
I will eventually refactor this into a dedicated `useWidgetState` custom hook. But doing the refactor for chart selections is a bit too much overhead and risk of breaking something. So I think its good to leave it as is and do some cleanup after the deadlines. There is anyways a lot to touche and refactor in the context of the lazy-loading support project.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+        )
+      }
+    }),
+    [widgetMgr, element]
```
Does `fragmentId` need to be added to the dependency array?

lukasmasuch replied:
yep, added it to dependencies 👍

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -271,41 +323,33 @@ function DataFrame({
-        let currentWidgetState = widgetMgr.getStringValue(
-          element as WidgetInfo
+      // Only update if there is actually a difference between editing and widget state
+      if (currentEditingState !== currentWidgetState) {
+        widgetMgr.setStringValue(
+          element as WidgetInfo,
+          currentEditingState,
+          {
+            fromUi: true,
+          },
+          fragmentId
```
This block seems to exist a couple of times, e.g. [here](https://github.com/streamlit/streamlit/pull/8411/files#diff-eda1f0faafab9bbafe2c7b8b34241c35192488443ab3e02f0291d91e40ef632fR269-R282)

Maybe it'd make sense to move it to an extra function

```typescript
function setDataFrameWidgetState(widgetMgr, element, oldState, newState, fragmentId) {
  if (newState === undefined || oldState === newState) {
    return
  }
  widgetMgr.setStringValue(
    element as WidgetInfo,
    currentEditingState,
    {
      fromUi: true,
    },
    fragmentId);
}
```
and call this function from the different places. This would help to understand that the logic is the same at all places and would probably make it a one-liner at the respective places

lukasmasuch replied:
I will eventually refactor this into a dedicated `useWidgetState` hook. But doing the refactor for chart selections is a bit too much overhead and risk of breaking something. So I think its good to leave it as is and do some cleanup after the deadlines. There is anyways a lot to touche and refactor in the context of the lazy-loading support project.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+      // state and the selection state.
+
+      const selectionState = {
+        // We use snake case here since this is the widget state
+        // that is sent and used in the backend. Therefore, it should
+        // conform with the Python naming conventions.
```
nit: the fields look like snake_case & camelcase would be the same for them since they are single words. Maybe if would make sense to rewrite it slightly to make this clear: 
`If you intend to rename the columns, remember to use snake_case since ...`

lukasmasuch replied:
I removed that part for now.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -385,7 +434,27 @@ function DataFrame({
   const isDynamicAndEditable =
     !isEmptyTable && element.editingMode === DYNAMIC && !disabled
+
+  const isRowSelectionActivated =
+    !isEmptyTable &&
+    !disabled &&
+    (element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_ROW) ||
+      element.selectionMode.includes(ArrowProto.SelectionMode.SINGLE_ROW))
+  const isMultiRowSelectionActivated =
+    isRowSelectionActivated &&
+    element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_ROW)
+
+  const isColumnSelectionActivated =
+    !isEmptyTable &&
+    !disabled &&
+    (element.selectionMode.includes(ArrowProto.SelectionMode.SINGLE_COLUMN) ||
+      element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_COLUMN))
+  const isMultiColumnSelectionActivated =
+    isColumnSelectionActivated &&
+    element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_COLUMN)
```
suggestion: I personally would move this into a convenience function outside of the component-function

```typescript
function getSelectionModes(isEmptyTable: boolean, isDisabled: boolean, element): [boolean, boolean, boolean, boolean] {
  const isSelectionActivated = !isEmptyTable && !disabled;

  if (!isSelectionActivated) {
    return;
  }

    const isMultiRowSelectionActivated = element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_ROW);
  const isRowSelectionActivated = isMultiRowSelectionActivated ||
      element.selectionMode.includes(ArrowProto.SelectionMode.SINGLE_ROW);
  
  const isMultiColumnSelectionActivated = element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_COLUMN);
  const isColumnSelectionActivated = isMultiColumnSelectionActivated || element.selectionMode.includes(ArrowProto.SelectionMode.SINGLE_COLUMN);

  return [isRowSelectionActivated, isMultiRowSelectionActivated, isColumnSelectionActivated, isMultiColumnSelectionActivated];
}
```

and then in the component-function it becomes a one-liner:

```typescript
const [isRowSelectionActivated, isMultiRowSelectionActivated, isColumnSelectionActivated, isMultiColumnSelectionActivated] = getSelectionModes(isEmptyTable, isDisabled, element);
```

raethlein replied:
Also if this is in a separate function, it is unit-testable 🙂

lukasmasuch replied:
I moved this out into a dedicated custom hook

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -482,12 +551,25 @@ function DataFrame({
         onCollapse={collapse}
         target={StyledResizableContainer}
       >
+        {((isRowSelectionActivated && isRowSelected) ||
+          (isColumnSelectionActivated && isColumnSelected)) && (
+          <ToolbarAction
+            label={"Clear selection"}
+            icon={Close}
+            onClick={() => {
+              clearSelection()
+              clearTooltip()
+            }}
+          />
+        )}
```
suggestion: I would probably move this further up:

```typescript
const includeToolbarAction = (isRowSelectionActivated && isRowSelected) ||
          (isColumnSelectionActivated && isColumnSelected);
const toolbarActionElement = includeToolbarAction && 
          <ToolbarAction
            label={"Clear selection"}
            icon={Close}
            onClick={() => {
              clearSelection()
              clearTooltip()
            }}
          />;
```
and then here just `{toolbarActionElement}` or maybe at least the boolean check and then

```typescript
{
includeToolbarAction && 
<ToolbarAction
  label={"Clear selection"}
  icon={Close}
  onClick={() => {
    clearSelection()
    clearTooltip()
  }}
/>
}
```

The whole component is already massive and for readability I think we should break the parts up a little bit

lukasmasuch replied:
I'm not sure if this will help a lot with the readability of moving this one item to some other place, I think its a bit more explicit the way it is right now. But I added a comment for some clarity.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -623,11 +705,19 @@ function DataFrame({
-            // Deactivate sorting for empty state and for large dataframes:
-            isEmptyTable || isLargeTable ? undefined : sortColumn
-          }
+          onHeaderClicked={(colIndex: number, _event) => {
```
is it possible to type the `_event` argument, e.g. `MouseEvent` or something like that?

lukasmasuch replied:
Nope, the type is glide-internal and not exported. But from typescript side, this is anyways typed

raethlein replied:
ah ok, thanks! is there an open issue to export the types or should we create one? I guess its good-practice to export types for public APIs

lukasmasuch replied:
Nope, I don't think its tracked currently. I can add that after the release

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -636,7 +726,65 @@ function DataFrame({
+
+              let updatedSelection = newSelection
+              if (
+                (isRowSelectionActivated || isColumnSelectionActivated) &&
+                newSelection.current !== undefined
+              ) {
+                // The default behavior is that row selections are cleared when a cell is selected.
+                // This is not desired when row selection is activated. Instead, we want to keep the
+                // row selection and only update the cell selection.
+                updatedSelection = {
+                  ...newSelection,
+                  rows: gridSelection.rows,
+                  columns: gridSelection.columns,
+                }
+              }
+
+              if (
+                rowSelectionChanged &&
+                newSelection.rows.length > 0 &&
+                columnSelectionChanged &&
+                newSelection.columns.length === 0
+              ) {
+                // Keep the column selection if row selection was changed
+                updatedSelection = {
+                  ...updatedSelection,
+                  columns: gridSelection.columns,
+                }
+              }
+              if (
+                columnSelectionChanged &&
+                newSelection.columns.length > 0 &&
+                rowSelectionChanged &&
+                newSelection.rows.length === 0
+              ) {
+                // Keep the row selection if column selection was changed
+                updatedSelection = {
+                  ...updatedSelection,
+                  rows: gridSelection.rows,
+                }
+              }
```
suggestion: I would move this into a function outside of the component function

```typescript
function getUpdatedSelection(existingSelection: GridSelection, newSelection: GridSelection): GridSelection {
    const rowSelectionChanged = !isEqual(
    newSelection.rows.toArray(),
    gridSelection.rows.toArray()
  )

  const columnSelectionChanged = !isEqual(
    newSelection.columns.toArray(),
    gridSelection.columns.toArray()
  )

  let updatedSelection = newSelection
  if (
    (isRowSelectionActivated || isColumnSelectionActivated) &&
    newSelection.current !== undefined
  ) {
    // The default behavior is that row selections are cleared when a cell is selected.
    // This is not desired when row selection is activated. Instead, we want to keep the
    // row selection and only update the cell selection.
    updatedSelection = {
      ...newSelection,
      rows: gridSelection.rows,
      columns: gridSelection.columns,
    }
  }

  if (
    rowSelectionChanged &&
    newSelection.rows.length > 0 &&
    columnSelectionChanged &&
    newSelection.columns.length === 0
  ) {
    // Keep the column selection if row selection was changed
    updatedSelection = {
      ...updatedSelection,
      columns: gridSelection.columns,
    }
  }
  if (
    columnSelectionChanged &&
    newSelection.columns.length > 0 &&
    rowSelectionChanged &&
    newSelection.rows.length === 0
  ) {
    // Keep the row selection if column selection was changed
    updatedSelection = {
      ...updatedSelection,
      rows: gridSelection.rows,
    }
  }

  return updatedSelection;
}
```

Then its also easier to write unit tests for it (same for the other comments in this regard)

lukasmasuch replied:
Unforuntalty, its not as simple. The `getUpdatedSelection` would also need `isRowSelectionActivated` and `isColumnSelectionActivated` as parameters + the `rowSelectionChanged` and `columnSelectionChanged` are required for the `applySelections` below :(

lukasmasuch replied:
Migrated the logic into a dedicated custom hook instead.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -636,7 +726,65 @@ function DataFrame({
+                (isRowSelectionActivated && rowSelectionChanged) ||
+                (isColumnSelectionActivated && columnSelectionChanged)
+              ) {
+                applySelections(updatedSelection)
```
nit: to make it clearer what this function is doing vs `setGridSelection`, we could call it something like `syncWithBackend` or something like that

lukasmasuch replied:
`syncWithBackend` might also a bit misleading since there are a couple of situations where it doesn't sync with backend. E.g. in form or if nothing has changed. I renamed it to `synceSelectionState`

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -703,18 +873,27 @@ function DataFrame({
-              rowMarkerTheme: {
-                bgCell: theme.bgHeader,
-                bgCellMedium: theme.bgHeader,
+              rowMarkers: {
+                kind: "checkbox",
+                checkboxStyle: "square",
+                theme: {
+                  bgCell: theme.bgHeader,
+                  bgCellMedium: theme.bgHeader,
+                },
```
is this a duplicate of [these lines](https://github.com/streamlit/streamlit/pull/8411/files#diff-eda1f0faafab9bbafe2c7b8b34241c35192488443ab3e02f0291d91e40ef632fR835-R842)? Is there a way to move it to a variable like

```typescript
const rowMarkers = (isRowSelectionActivated || element.editingMode === DYNAMIC) && {
    rowMarkers: {
      kind: "checkbox",
      checkboxStyle: "square",
      theme: {
        bgCell: theme.bgHeader,
        bgCellMedium: theme.bgHeader,
    },
}
```

and apply it only once here?

lukasmasuch replied:
Hmm, I think it is a bit of a coincidence that it's the same, but this might not be the case in the future. In a previous state of this PR, we already had differences, like using the rounded checkbox style in certain situations.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -703,18 +873,27 @@ function DataFrame({
               onHeaderClicked: undefined,
             })}
+          // Disable some features when the widget is disabled:
+          {...(disabled && {
+            rowSelect: "none",
+            columnSelect: "none",
+            onRowAppended: undefined,
```
Is it possible to move this into a variable instead of relying on the fact that these properties are set last and overwrite the previous ones?
Like

```typescript
const rowSelect = disabled ? "none" : "multi";
```

and then apply `rowSelect` only once? Its a little bit tough to look at the component and keep in mind what was already set, what is overwritten etc.

lukasmasuch replied:
That we the previous state, I actually just refactored this out to make it a bit simpler to read

lukasmasuch replied:
Would you prefer to have this moved back into the other parts?

lukasmasuch replied:
In some of the checks it would look like this: 
```
             columnSelect: disabled
                ? "none"
                : isMultiColumnSelectionActivated
                ? "multi"
                : "single",
```

raethlein replied:
yeah, I personally would find that easier than having the `columnSelect` key multiple times overriding each other in different scenarios. This way we would see the value of that field in one place. But if you already had that and found it difficult, it might be more like a personal preference and not worth doing it. So with the arguments brought up, I will leave it to you for the final decision

lukasmasuch replied:
Ok, I move it up again.

raethlein commented on lib/streamlit/elements/arrow.py:
```diff
@@ -51,9 +72,112 @@
     None,
 ]
 
+SelectionMode: TypeAlias = Literal[
+    "single-row", "multi-row", "single-column", "multi-column"
+]
+_SELECTION_MODES: Final[set[SelectionMode]] = {
+    "single-row",
+    "multi-row",
+    "single-column",
+    "multi-column",
+}
```
curious: is this a python-typing thing that we have to define it twice like this?

lukasmasuch replied:
These are a bit different things the type alias is purely for type checking, while _SELECTION_MODES is a value that we can use within the python logic.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -262,6 +234,86 @@ function DataFrame({
+      // into a single function that updates the widget state with both the editing
+      // state and the selection state.
+
+      const selectionState = {
```
should we give this a type, like we do with `DataFrameState` on the Python-side?

lukasmasuch replied:
Good point, I added an interface for this

raethlein commented on lib/streamlit/elements/arrow.py:
````diff
@@ -119,6 +280,30 @@ def dataframe(
+            - "ignore" (default): Streamlit will not react to any selection events in the chart.
+            - "rerun": Streamlit will rerun the app when the user selects rows or columns in the table.
+              In this case, ```st.dataframe``` will return the selection data as a dictionary.
+            - callable: If a callable is provided, Streamlit will rerun and execute the callable as a
````
nit: maybe worth to mention if the callback accepts any parameters (probably not right?)

lukasmasuch replied:
Nope, doesn't accept any parameters. But thats probably something for Debbie to unify since it uses the same text as other comparable instances.

raethlein commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -53,24 +53,25 @@
     {
         "button": "trigger_value",
-        "download_button": "trigger_value",
+        "camera_input": "file_uploader_state_value",
```
❤️

lukasmasuch committed:
Move selection logic into dedicated hook

lukasmasuch committed:
Fix comment

lukasmasuch committed:
Some refactoring

lukasmasuch committed:
Add comment

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Apply some fixes

lukasmasuch committed:
Update test

lukasmasuch committed:
Merge branch 'develop' into feature/dataframe-selections

lukasmasuch committed:
Add custom hook unit tests

lukasmasuch committed:
Merge branch 'feature/dataframe-selections' of https://github.com/streamlit/streamlit into feature/dataframe-selections

lukasmasuch committed:
Fix imports

lukasmasuch committed:
Add more unit tests

raethlein committed:
Add e2e tests for the dataframe selections (#8634)

## Describe your changes

Adds e2e tests for the dataframe row and column selections introduced in
this PR: https://github.com/streamlit/streamlit/pull/8411

The tests run fine on my local Mac but not on the CI/CD pipeline yet.
Likely something with the Meta key for multiple column selection.

## GitHub Issue Link (if applicable)

## Testing Plan

- see testing plan of https://github.com/streamlit/streamlit/pull/8411

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

lukasmasuch commented on e2e_playwright/st_dataframe_selections_test.py:
```diff
@@ -0,0 +1,310 @@
+
+def test_single_row_select(app: Page):
+    canvas = _get_single_row_select_df(app)
+    # bounding_box = canvas.bounding_box()
```
@raethlein nit: can this be removed?

raethlein replied:
yep!

lukasmasuch commented on e2e_playwright/st_dataframe_selections_test.py:
```diff
@@ -0,0 +1,310 @@
+    expect(selection_text).to_have_count(1)
+
+
+# def test_in_form_selection_and_session_state(app: Page):
```
@raethlein can this test case be removed?

raethlein replied:
yep! the same tests exists below it but in a fixed variant; this was a non-working state. So the commented out test can be removed

lukasmasuch commented on e2e_playwright/st_dataframe_selections_test.py:
```diff
@@ -0,0 +1,310 @@
+    selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+    expect(selection_text).to_have_count(1)
+
+    # submit the form. Without the timeout the test is flaky, presumably because the submit-button does not have an eventlistener on time :/
```
@raethlein It might be because of the 150ms debounce, maybe its better to use 200ms to make sure that its guaranteed to be longer then the debounce.

lukasmasuch commented on e2e_playwright/st_dataframe_selections_test.py:
```diff
@@ -0,0 +1,310 @@
+    canvas = _get_multi_row_and_column_select_df(app)
+
+    # toolbar has three buttons: download, search, fullscreen
+    data_editor_toolbar = canvas.get_by_test_id("stElementToolbar")
```
@raethlein 
nit: `dataframe_toolbar`

lukasmasuch commented on e2e_playwright/st_dataframe_selections_test.py:
```diff
@@ -0,0 +1,310 @@
+    _select_some_rows_and_columns(app, canvas)
+    _expect_multi_row_multi_column_selection(app)
+
+    assert_snapshot(canvas, name="st_data_editor-multi_row_multi_column_selection")
```
@raethlein 
nit: this needs to be `st_dataframe-` since the data editor does not support selections at the moment

raethlein replied:
💡 gotcha

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -371,6 +451,10 @@ function DataFrame({
+    }
+
     const formClearHelper = new FormClearHelper()
     formClearHelper.manageFormClearListener(
```
TODO: we need to clear also the selection state here.

lukasmasuch replied:
Fixed

raethlein committed:
Address e2e tests feedback

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/hooks/useSelectionHandler.ts:
```diff
@@ -0,0 +1,195 @@
+  const isColumnSelected = gridSelection.columns.length > 0
+  const isCellSelected = gridSelection.current !== undefined
+
+  const processSelectionChange = React.useCallback(
```
TODO: Add a comment to this callack

lukasmasuch commented on frontend/lib/src/components/widgets/DataFrame/hooks/useSelectionHandler.ts:
```diff
@@ -0,0 +1,195 @@
+
+import { Arrow as ArrowProto } from "@streamlit/lib/src/proto"
+
+export type SelectionHandlerReturn = {
```
TODO: add some comments here

raethlein replied:
one comment to add might be why we have `clearCellSelection` as a distinct clearer but do not differentiate between row & column selection

lukasmasuch replied:
Removed clearCellSelection and just added parameters to clearSelections to keep row & column selections

raethlein committed:
Update variable name

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+  const syncSelectionState = React.useCallback(
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, (newSelection: GridSelection) => {
+      // If we want to support selections also with in the editable mode,
```
nit: "with in" -> "within"

lukasmasuch replied:
Updated

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, (newSelection: GridSelection) => {
+      // If we want to support selections also with in the editable mode,
+      // we would need to integrate the `applyEdits` and `syncSelections` functions
```
nit: should be "applyEdits" -> "syncEditStates" right?

lukasmasuch replied:
Updated

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+      selectionState.select.columns = newSelection.columns
+        .toArray()
+        .map(columnIdx => {
+          return getColumnName(originalColumns[columnIdx])
```
curious: why are we using the index for rows but the names for columns? Should the reason be documented somewhere (maybe it is and I missed it)?

lukasmasuch replied:
We are using it here since we decided to use the same identifying attributes for edits with data editor. And the general decision is to use what is most descriptive and easiest to use in the python backend. For columns that's the name, for rows we don't really have something equivalent. We evaluated to use the row index (index columns),  but there is a lot of complexity around that ... and most dataframes don't even have an index defined.

lukasmasuch replied:
Added a comment about this to the interface

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+        currentWidgetState !== newWidgetState
+      ) {
+        widgetMgr.setStringValue(
+          {
+            id: element.id,
+            formId: element.formId,
+          } as WidgetInfo,
```
the pattern

```typescript
{
  id: element.id,
  formId: element.formId,
} as WidgetInfo
```

seems to be a couple of times in the changes here. It might be a good idea to have a convenience function (perhaps in the `WidgetMgr`) to generate the right object. E.g.:

```typescript
function getWidgetMgrKey(element: Element): WidgetInfo {
  return { id: element.id, formId: element.formId } as WidgetInfo
}
```

and then call that function here. This might also help to not forget including the `formId`

lukasmasuch replied:
yeah, I actually split it up from the current way of just passing in the element here `element as WidgetInfo` as we do in most other widgets. Reason was that it makes it easier to correctly manage the effect dependencies. What we should do here is to move `formId` into its own parameter and treat it the same way as fragmentID. But that refactoring needs to happen in a dedicated PR. So I don't want to introduce another method around that right now, and instead just correctly refactor this.

lukasmasuch replied:
I added a tech debt icon [here](https://www.notion.so/snowflake-corp/Refactor-formId-into-its-own-property-for-all-widget-state-methods-59b5768ec45c4a5a84664b27b75ad3bd)

raethlein replied:
gotcha, thanks for the context!

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+  const refreshCells = React.useCallback(
+    (
+      cells: {
+        cell: [number, number]
```
nit: perhaps we should have a type for this that has a comment; I assume the `[number, number]` is the row & column index?

lukasmasuch replied:
Changed it to use the relevant type that is exposed by glide-data-grid

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+      element.id,
+      element.formId,
+      widgetMgr,
+      element,
```
do we need to pass `element` here, since we already pass `element.id` and `element.formid` and we don't seem to use anything else from element inside the function

lukasmasuch replied:
good point, yep `element` needs to be removed. `element` is anyways not working correctly as dependency since its an object that might change its identity.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +241,150 @@ function DataFrame({
+      widgetMgr,
+      element,
+      fragmentId,
+      numRows,
```
`numRows` does not seem to be directly used inside the function. Can it be removed or is it required to refresh the function anyways? If yes, a comment might be useful to say why

lukasmasuch replied:
yep, can be removed. Its an historical artifact since the `updateNumRows` was included here before.

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -426,6 +508,8 @@ function DataFrame({
   React.useEffect(() => {
     // Clear cell selections if fullscreen mode changes
     clearCellSelection()
+    // Only run this on changes to the fullscreen mode:
+    // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [isFullScreen])
```
Could you add a comment why this is the case please? Its not directly obvious too me why full-screen mode should impact what cells are selected

lukasmasuch replied:
This is from a long time ago, but there were some weird side effects if the fullscreen mode was changed and cells were selected. Since it doesn't really add value to keep the cell selection, I just cleaned it to avoid dealing with these side effects.

lukasmasuch replied:
i don't have a lot more specific details on that, would need to test around again

raethlein replied:
alrighty. maybe it's a good idea to add that^ as a brief comment:
`// In the past we saw some weird side-effects, so we decided to clean it when entering fullscreen-mode. If we decide to change this, we have to play around and get to the bottom of it`

lukasmasuch replied:
added 👍

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -623,20 +724,32 @@ function DataFrame({
+              return undefined
+            }
+            if (isRowSelectionActivated && isRowSelected) {
+              // Row selections and column sorting is not compatible.
```
Could you elaborate here briefly why its not compatible and what the delta would be to make it compatible? I assume its because we would need to map the old indices to the new indices and update the selection array with the mapped indices.

lukasmasuch replied:
Added a comment 👍

raethlein commented on frontend/lib/src/components/widgets/DataFrame/hooks/useSelectionHandler.ts:
```diff
@@ -0,0 +1,195 @@
+      gridSelection,
+      isRowSelectionActivated,
+      isColumnSelectionActivated,
+      syncSelectionState,
```
do we have to pass `setGridSelection` here, too? It's a state-setter so probably not as it usually does not change, but want to sanity check

lukasmasuch replied:
Nope, afaik state-setter methods don't need to be included in dependencies and eslint also seems to be fine with this.

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Merge branch 'feature/dataframe-selections' of https://github.com/streamlit/streamlit into feature/dataframe-selections

lukasmasuch committed:
Add comment

lukasmasuch committed:
Use glide type

lukasmasuch committed:
Rename

lukasmasuch committed:
Update comment

lukasmasuch committed:
Update comment

lukasmasuch committed:
Added comment

lukasmasuch committed:
Removed not required dependencies

lukasmasuch committed:
Add comment

lukasmasuch committed:
Fix clear on submit for form

lukasmasuch committed:
Remove unused import

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Remove clearCellSelections callback

lukasmasuch committed:
Add comment

lukasmasuch committed:
Move the disabled checks up

lukasmasuch committed:
Update glide data grid to alpha 8

lukasmasuch committed:
Add reconstruction of selection state

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/dataframe-selections

lukasmasuch committed:
Additional improvements

lukasmasuch committed:
Update snapshots

lukasmasuch committed:
Update snapshot

lukasmasuch committed:
Skip firefox

lukasmasuch committed:
Add comment

raethlein committed:
Improve e2e-test text-expectation check

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -233,13 +213,18 @@ function DataFrame({
+  /**
+   * On the first rendering, try to load the initial editing state
+   * from widget state if it exists. This is required in the case
+   * that other elements are inserted before this widget.
```
Could you elaborate a little bit how other elements can impact the editing state of a dataframe widget?

lukasmasuch replied:
Added more to the comment. TLDR: if elements are inserted before, it can happen that the components gets unmounted and loses all state.

lukasmasuch replied:
btw. this is the case (problem) for all widgets and even non-widgets e.g. tabs or expander can lose the state as well because of this

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -263,50 +248,213 @@ function DataFrame({
+        formId: element.formId,
+      } as WidgetInfo)
+
+      if (initialWidgetValue) {
```
nit: in order to reduce nesting and thus increase readability, it looks like you could do the reverse check:

```typescript
if (!initialWidgetValue) {
  return
}
```

lukasmasuch replied:
Updated 👍

raethlein commented on frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx:
There is a lot of new / updated logic in this file but no new unit tests are added to `DataFrame.test.tsx`. Do you think that e2e tests alone are sufficient?

lukasmasuch replied:
The aspects that got added are quite complicated to test with RTL and it would also mostly do the same as our e2e tests, but in a less complete way. Most of the new logic in Dataframe.tsx is related to widget state handling. I plan to migrate this logic into a dedicated custom hook -  e.g. useWidgetState - which would also allow better testing. But its something that will require a dedicated follow-up PR.

raethlein reviewed (approved):
🚀

lukasmasuch committed:
Add test for usage in fragment

lukasmasuch committed:
Fix issue with selections on cell selection

lukasmasuch committed:
Remove console log

lukasmasuch committed:
Fix e2e test

lukasmasuch committed:
Clean up tests
